File: Background-Subtraction/background_subtr_bgslib.py
import argparse

import cv2
import pybgs as bgs


def get_bgslib_result(video_to_process):
    # create VideoCapture object for further video processing
    captured_video = cv2.VideoCapture(video_to_process)
    # check video capture status
    if not captured_video.isOpened:
        print("Unable to open: " + video_to_process)
        exit(0)

    # instantiate background subtraction
    background_subtr_method = bgs.SuBSENSE()

    while True:
        # read video frames
        retval, frame = captured_video.read()

        # check whether the frames have been grabbed
        if not retval:
            break

        # resize video frames
        frame = cv2.resize(frame, (640, 360))

        # pass the frame to the background subtractor
        foreground_mask = background_subtr_method.apply(frame)
        # obtain the background without foreground mask
        img_bgmodel = background_subtr_method.getBackgroundModel()

        # show the current frame, foreground mask, subtracted result
        cv2.imshow("Initial Frames", frame)
        cv2.imshow("Foreground Masks", foreground_mask)
        cv2.imshow("Subtraction result", img_bgmodel)

        keyboard = cv2.waitKey(10)
        if keyboard == 27:
            break


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--input_video",
        type=str,
        help="Define the full input video path",
        default="space_traffic.mp4",
    )

    # parse script arguments
    args = parser.parse_args()

    # start BS-pipeline
    get_bgslib_result(args.input_video)

File: Background-Subtraction/background_subtr_bgslib.cpp
#include <iostream>
#include <algorithm>
#include <iterator>
#include <vector>
#include <opencv2/videoio.hpp>

#include "algorithms/algorithms.h"

using namespace cv;
using namespace std;

const char* input_params = "{ input | space_traffic.mp4 | Define the full input video path }";

void get_bgslib_result(String video_to_process)
{
    // create VideoCapture object for further video processing
    VideoCapture capture(samples::findFile(video_to_process));
    if (!capture.isOpened()) {
        //error in opening the video input
        cerr << "Unable to open: " << video_to_process << endl;
        return;
    }

    // instantiate background subtraction model
    SuBSENSE background_subtr_method = SuBSENSE();

    Mat frame, fgMask, background;
    while (true) {
        capture >> frame;

        // check whether the frames have been grabbed
        if (frame.empty())
            break;

        // resize video frames
        resize(frame, frame, Size(640, 360));

        try
        {
            // pass the frame to the model processor
            background_subtr_method.process(frame, fgMask, background);
        }
        catch (exception& e)
        {
            cout << "Exception occurred" << endl;
            cout << e.what() << endl;
        }

        // show the current frame, foreground mask, subtracted result
        imshow("Initial Frames", frame);
        imshow("Foreground Masks", fgMask);
        imshow("Subtraction Result", background);

        int keyboard = waitKey(10);
        if (keyboard == 27)
            break;
    }
}

int main(int argc, char* argv[])
{
    CommandLineParser parser(argc, argv, input_params);
    // start BS-pipeline
    get_bgslib_result(parser.get<String>("input"));

    return 0;
}